snuba/web/query.py

```diff
@@ -209,33 +268,69 @@ def log_query_and_update_stats(
     """
 
     stats.update(query_settings)
+
+    query_metadata.query_list.append(
+        ClickhouseQueryMetadata(
+            sql=sql, timing=timer.for_json(), stats=stats, status=status, trace_id=trace_id,
+        )
+    )
+
+    return stats
+
+
+def record_query(
+    request: Request, timer: Timer, query_metadata: SnubaQueryMetadata
+) -> None:
     if settings.RECORD_QUERIES:
         # send to redis
-        state.record_query(
-            {
-                "request": request.body,
-                "sql": sql,
-                "timing": timer,
-                "stats": stats,
-                "status": status,
-            }
-        )
+        query_metadata.http_referrer = http_request.referrer
+        state.record_query(query_metadata.to_dict())
```

Comment:
I think that `record_query` should be updated to take `SnubaQueryMetadata` here instead.

Reply:
I wanted to avoid record_query needing to know about the SnubaQueryMetadata class. Updated anyway.

Reply:
Meh, unless I misunderstand what you mean attempting to import the QueryMetadata type into state causes circular import issues

Reply:
This is wrong now:
> def record_query(query_metadata: SnubaQueryMetadata) -> None:

could you please run
> mypy --strict snuba | grep QueryMetadata 

(well without the grep would be useful as well but it will return like 500 results. I still think that a quick check to those results at least once per PR is useful)

Also, more important, all unit tests run with RECORD_QUERIES=False, you need to test this manually to ensure it does not break in prod.